pennylane/qnn/keras.py

```diff
@@ -0,0 +1,292 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module is for PennyLane-Keras integration."""
+import functools
+import inspect
+from collections.abc import Iterable
+from typing import Optional
+
+try:
+    import tensorflow as tf
+    from tensorflow.keras.layers import Layer
+    from pennylane.interfaces.tf import to_tf
+
+    CORRECT_TF_VERSION = int(tf.__version__.split(".")[0]) > 1
+except ImportError:
+    from abc import ABC
+
+    Layer = ABC
+    CORRECT_TF_VERSION = False
+
+from pennylane.qnodes import QNode
+
+
+class KerasLayer(Layer):
+    """A Keras Layer_ for integrating PennyLane QNodes with the Keras API.
+
+    Args:
+        qnode (qml.QNode): the PennyLane QNode to be converted into a Keras Layer_
+        weight_shapes (dict[str, tuple]): a dictionary mapping from all weights used in the QNode to
+            their corresponding shapes
+        output_dim (int): the output dimension of the QNode
+        weight_specs (dict[str, dict]): An optional dictionary for users to provide additional
+            specifications for weights used in the QNode, such as the method of parameter
+            initialization. This specification is provided as a dictionary with keys given by the
+            arguments of the `add_weight()
+            <https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer#add_weight>`__
+            method and values being the corresponding specification.
+        **kwargs: additional keyword arguments passed to the Layer_ base class
+
+    This class converts a :func:`~.QNode` to a Keras Layer_, which can be used within the Keras
+    `Sequential <https://www.tensorflow.org/api_docs/python/tf/keras/Sequential>`__ or
+    `Model <https://www.tensorflow.org/api_docs/python/tf/keras/Model>`__ classes for
+    creating quantum and hybrid models:
+
+    .. code-block:: python
+
+        qlayer = qml.qnn.KerasLayer(qnode, weight_shapes, output_dim=2)
+        model = tf.keras.models.Sequential([qlayer, tf.keras.layers.Dense(2)])
+
+    The signature of QNode must contain an ``inputs`` named argument for input data, with all
+    other arguments to be treated as internal weights. A valid ``qnode`` for the example
+    above would be:
+
+    .. code-block:: python
+
+        n_qubits = 2
+        dev = qml.device("default.qubit", wires=n_qubits)
+
+        @qml.qnode(dev)
+        def qnode(inputs, weights_0, weight_1):
+            qml.RX(inputs, wires=0)
+            qml.Rot(*weights_0, wires=0)
+            qml.RY(weight_1, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+    The internal weights of the QNode are automatically initialized within the
+    :class:`~.KerasLayer` and must have their shapes specified in a ``weight_shapes`` dictionary.
+    For example:
+
+    .. code-block::
+
+        weight_shapes = {"weights_0": 3, "weight_1": 1}
+
+    **Example**
```

Comment:
I would say that the code blocks _above_ constitute the basic example, whereas the example below with full training and templates should go within the `.. UsageDetails::` section? Could be fixed by just moving `**Example**` to line 55, and moving `.. UsageDetails::` to here.

Reply:
Especially because the example below is not a minimal example, as it uses scikit learn datasets etc.

Reply:
I agree!